nntrainer/models/neuralnet.cpp

```diff
@@ -162,31 +164,71 @@ int NeuralNetwork::initialize() {
  */
 NeuralNetwork::~NeuralNetwork() { model_graph.reset(); }
 
-void NeuralNetwork::setLabels(sharedConstTensors label) {
-  auto fill_label = [&label](auto const &layer_node) {
-    NNTR_THROW_IF(label.size() != layer_node->getNumOutputs(),
-                  std::invalid_argument)
-      << "label size does not match with the layer requirements"
-      << " layer: " << layer_node->getName() << " label size: " << label.size()
-      << " requirements size: " << layer_node->getNumOutputs();
+static void setLabels(const std::vector<Tensor> &data,
+                      const std::vector<Var_Grad *> &label_list) {
 
-    for (unsigned int i = 0; i < layer_node->getNumOutputs(); i++) {
-      layer_node->getOutputGradUnsafe(i) = *label[i];
-    }
-  };
+  NNTR_THROW_IF(!data.empty() && data.size() != label_list.size(),
+                std::invalid_argument)
+    << "label size does not match with the network requirements"
+    << " label size: " << data.size()
+    << " requirements size: " << label_list.size();
 
-  auto clear_label = [](auto const &layer_node) {
-    for (unsigned int i = 0; i < layer_node->getNumOutputs(); i++) {
-      layer_node->getOutputGradUnsafe(i) = Tensor();
-    }
-  };
+  /// feed or clear label
+  for (unsigned int idx = 0; idx < data.size(); idx++) {
+    if (data.empty())
+      label_list[idx]->initializeGradient();
+    else if (data.size() == 1)
+      label_list[idx]->initializeGradient(data[0]);
+    else
+      label_list[idx]->initializeGradient(data[idx]);
+  }
+}
+
+static void setInputs(const std::vector<Tensor> &data,
+                      const std::vector<Var_Grad *> &input_list) {
+
+  NNTR_THROW_IF(data.size() != input_list.size(), std::invalid_argument)
+    << "input size does not match with the network requirements"
+    << " input size: " << data.size()
+    << " requirements size: " << input_list.size();
 
   /// feed or clear label
-  for (auto iter = model_graph.cbegin(); iter != model_graph.cend(); iter++) {
-    auto const &lnode = *iter;
-    if (lnode->requireLabel()) {
-      label.empty() ? clear_label(lnode) : fill_label(lnode);
-    }
+  for (unsigned int idx = 0; idx < input_list.size(); idx++) {
+    input_list[idx]->initializeVariable(data[idx]);
+  }
+}
+
+static void setLabels(sharedConstTensors &data,
```

Comment:
There is another setLabels above, one of them should depend on the other one?

Reply:
Yeah, its done with #1544. I will cherry-pick that change here.

Reply:
done.